saleor/order/migrations/tasks/saleor3_9.py

```diff
@@ -0,0 +1,137 @@
+    if line_net == line_gross and all([line_net != 0, line_gross != 0]):
+        line_net, line_gross = add_taxes_to_price(
+            line_net, tax_rate, prices_entered_with_tax
+        )
+        setattr(
+            line, f"{field_name}_net_amount", quantize_price(line_net, line.currency)
+        )
+        setattr(
+            line,
+            f"{field_name}_gross_amount",
+            quantize_price(line_gross, line.currency),
+        )
+        lines_to_update.add(line)
+
+
+def recalculate_tax_for_prices(
+    line, prices_entered_with_tax, tax_rate, lines_to_update
+):
+    for field in ["undiscounted_unit_price", "undiscounted_total_price"]:
+        update_fields_for_line(
+            field, line, tax_rate, prices_entered_with_tax, lines_to_update
+        )
+
+
+def get_order_prices_entered_with_tax_mapping(
+    order_lines, order_model, channel_model, tax_configuration
+):
+    orders = order_model.objects.filter(id__in=order_lines.values("order_id"))
+    channels = channel_model.objects.filter(id__in=orders.values("channel_id"))
+    tax_configurations = tax_configuration.objects.filter(
+        id__in=channels.values("tax_configuration__id")
+    )
+    tax_conf_to_prices_entered_with_taxes = {
+        tc_id: price_entered_with_taxes
+        for tc_id, price_entered_with_taxes in tax_configurations.values_list(
+            "id", "prices_entered_with_tax"
+        )
+    }
+
+    channel_to_prices_entered_with_taxes = dict()
+    for channel_id, tax_conf_id in channels.values_list("id", "tax_configuration__id"):
+        channel_to_prices_entered_with_taxes[
+            channel_id
+        ] = tax_conf_to_prices_entered_with_taxes[tax_conf_id]
+
+    order_to_prices_entered_with_taxes = dict()
+    for order_id, channel_id in orders.values_list("id", "channel__id"):
+        order_to_prices_entered_with_taxes[
+            order_id
+        ] = channel_to_prices_entered_with_taxes[channel_id]
+
+    return order_to_prices_entered_with_taxes
+
+
+@app.task
+def recalculate_undiscounted_prices():
+    order_lines = OrderLine.objects.filter(
+        (
+            Q(
+                undiscounted_unit_price_net_amount=F(
+                    "undiscounted_unit_price_gross_amount"
+                )
+            )
+            | Q(
+                undiscounted_total_price_net_amount=F(
+                    "undiscounted_total_price_gross_amount"
+                )
+            )
+        )
+        & Q(tax_rate__gt=0)
+    ).order_by("-created_at")[:BATCH_SIZE]
+
+    if not order_lines:
+        task_logger.info("No order lines to update.")
+        return
+
+    order_to_prices_with_tax_map = get_order_prices_entered_with_tax_mapping(
+        order_lines,
+        order_model=Order,
+        channel_model=Channel,
+        tax_configuration=TaxConfiguration,
+    )
+
+    lines_to_update = set()
+    for line in order_lines:
+        prices_entered_with_tax = order_to_prices_with_tax_map[line.order.id]
```

Comment:
Have you tested the celery task with database min 1 mln of orders?
Also, you're calling`line.order.id` without prefetch.